canonical/cos-configuration-k8s-operator
# fix: error on stop

## Issue
When removing cos-config via `juju remove-application`, the charm goes into **error** state and is not removed.

The error is the following:
```
...
  File "./src/charm.py", line 483, in tracing_endpoint
    if self._tracing.is_ready():
AttributeError: 'COSConfigCharm' object has no attribute '_tracing'
```

This happens because the charm `__init__` returns early on removal if the storage is detached ([code](https://github.com/canonical/cos-configuration-k8s-operator/blob/235ad09a3e57c91a0554ab5f1bf0485594aab444/src/charm.py#L89-L92)):
```python
... # in __init__
if len(self.model.storages["content-from-git"]) == 0:
    # Storage isn't available yet. Since storage becomes available early enough, no need
    # to observe storage-attached and complicate things; simply abort until it is ready.
    return
...
```

Then, the call to `self._tracing.is_ready()` which happens [here](https://github.com/canonical/cos-configuration-k8s-operator/blob/235ad09a3e57c91a0554ab5f1bf0485594aab444/src/charm.py#L483) (and is called by the `@trace_charm` decorator) fails because `self._tracing` is not defined.


## Solution

Move the `self._tracing` definition above the early constructor return.


## Testing Instructions
1. `charmcraft pack` (from the PR)
2. `juju deploy ./the.charm cos-config`
3. `juju remove-application --destroy-storage cos-config`
4. Verify the application is removed correctly and doesn't error on stop
